### PR TITLE
Deprioritize HEVC-over-FLV formats

### PR DESCRIPTION
**Pull request by Lesmiscore**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

</details>

Looks like BiliBili is offering HEVC-over-FLV formats, which isn't in standards and even not supported by ffmpeg. ([How do browsers play? There's a library to demux for them](https://github.com/xqq/mpegts.js/blob/6bf10e8a5be96543aaf3fb4bb7dcfcf8e7a21f02/src/demux/flv-demuxer.js#L843))

Although they could have been skipped, I think they should be available for download for anyone interested in.

ref. https://trac.ffmpeg.org/ticket/6389
ref. #5821

... wait, should we implement some scheme to deprioritize such formats (i.e. impossible or out-of-spec combination of container/codecs) globally? @pukkandan 

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [ ] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [x] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

#5821

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

### CODE REVIEW

**Comment by pukkandan** [comment]
> should we implement some scheme to deprioritize such formats (i.e. impossible or out-of-spec combination of container/codecs) for all the formats?

You mean in core code? We could add this logic here https://github.com/yt-dlp/yt-dlp/blob/c73355510629e3eda5a79d4e2876a35316ca6ed2/yt_dlp/utils.py#L6307-L6309

**Comment by Lesmiscore** [comment]
@pukkandan Maybe close this too and move the logic there?

**Comment by pukkandan** [comment]
Sounds good

File: yt_dlp/utils.py [hunk 1]
```diff
@@ -6307,6 +6307,12 @@ def calculate_preference(self, format):
         # if format.get('preference') is None and format.get('ext') in ('f4f', 'f4m'):  # Not supported?
         #    format['preference'] = -1000
 
+        if format.get('preference') is None and format.get('ext') == 'flv' and format.get('vcodec') in ('hevc', 'h265'):
+            # HEVC-over-FLV is out-of-spec by FLV's original spec
+            # ref. https://trac.ffmpeg.org/ticket/6389
+            # ref. https://github.com/yt-dlp/yt-dlp/pull/5821
+            format['preference'] = -1000
```
Comment (pukkandan) [inline yt_dlp/utils.py:6314]: Note: -1000 means the format won't be visible in `-F`

Comment (Lesmiscore) [reply yt_dlp/utils.py:6314]:
```
$ ./yt-dlp.sh -F https://live.bilibili.com/6 
[BiliLive] Extracting URL: https://live.bilibili.com/6
[BiliLive] 6: Downloading JSON metadata
[BiliLive] 6: Downloading JSON metadata
[BiliLive] 6: Downloading JSON metadata
[BiliLive] 6: Downloading JSON metadata
[BiliLive] 6: Downloading JSON metadata
[BiliLive] 6: Downloading JSON metadata
[BiliLive] 6: Downloading JSON metadata
[BiliLive] 6: Downloading JSON metadata
[info] Available formats for 6:
ID               EXT RESOLUTION │ PROTO │ VCODEC ACODEC  MORE INFO
──────────────────────────────────────────────────────────────────
ultra_high_res-0 flv unknown    │ https │ hevc   unknown 超清
ultra_high_res-1 flv unknown    │ https │ hevc   unknown 超清
source-0         flv unknown    │ https │ hevc   unknown 原画
source-1         flv unknown    │ https │ hevc   unknown 原画
high_res-0       flv unknown    │ https │ avc    unknown 高清
high_res-1       flv unknown    │ https │ avc    unknown 高清
ultra_high_res-2 flv unknown    │ https │ avc    unknown 超清
ultra_high_res-3 flv unknown    │ https │ avc    unknown 超清
source-2         flv unknown    │ https │ avc    unknown 原画
source-3         flv unknown    │ https │ avc    unknown 原画
```
I do see it now - an another bug?

Comment (pukkandan) [reply yt_dlp/utils.py]: hm.. -1000 itself is shown, apparently. https://github.com/yt-dlp/yt-dlp/blob/c73355510629e3eda5a79d4e2876a35316ca6ed2/yt_dlp/YoutubeDL.py#L3623

I don't know whether this is intentional or a bug. I actually don't know of any extractor that uses this feature or even why this exists...

Comment (Lesmiscore) [reply yt_dlp/utils.py]: Anyway I changed it to -100 since there's no point in forbidding from downloading

**Comment by Vangelis66** [comment]
> HEVC-over-FLV formats, which isn't in standards and even **not supported by ffmpeg.**

FWIW, there's (open-source?) code in GitHub (from Oct 2020) which implements exactly that (**HEVC inside FLV container**): 
https://github.com/numberwolf/FFmpeg-QuQi-H265-FLV-RTMP

Another dev patched a slightly newer ffmpeg version from Nov 2020, see [here](https://github.com/numberwolf/FFmpeg-QuQi-H265-FLV-RTMP/issues/1#issuecomment-731303754); hopefully, current ffmpeg 5.x.x could be patched (and compiled?) in a similar fashion 😜 ...

Source: [SO](https://stackoverflow.com/questions/65317708/i-can-t-play-hevc-encoded-flv-format-video-files-with-ffplay)

**Comment by Lesmiscore** [comment]
https://trac.ffmpeg.org/ticket/6389#comment:1
They doesn't seem to support it in master, although it doesn't stop us from patching ffmpeg to allow HEVC-over-FLV

**Comment by Lesmiscore** [comment]
https://github.com/yt-dlp/FFmpeg-Builds/commit/6561554a07f6476ebefcc1b7967a5e80efee5681

**Comment by nihil-admirari** [comment]
TLDR:
1. A [patch](https://github.com/yt-dlp/FFmpeg-Builds/blob/master/patches/ffmpeg/master/0002-Support-for-HEVC-over-FLV.patch) was added to FFmpeg-Builds.
2. [GMP site](https://gmplib.org)'s certificate has expired, and the owners are aware of it:

   > YES, THE CERTIFICATE IS EXPIRED. Sorry about that. We're working on the renewal process.

   Builds are currently failing. I've managed to build [a patched FFmpeg](https://github.com/nihil-admirari/FFmpeg-Builds/releases) before the certificate expired, but it uses an outdated version of the patch and it's only for Linux.
2. It is completely __untested__, since I don't have premium access to BiliBili, and I only get
   ```
   > yt-dlp https://www.bilibili.com/video/BV1Dx411n7Pg?p=32
   ...
   [BiliBili] Format(s) 1080P 高清, 360P 流畅, 480P 清晰, 720P 高清 are missing; you have to login or become premium member to download them
   ```
3. Support for VP8, VP9, AV1, and OPUS was omitted because:
   a. It wasn't requested.
   b. It causes too many questions.
4. `yt-dlp` should probably warn users and remux HEVC-over-FLC upon encountering, to prevent playback issues.

<hr>

Gory details:

> Another dev patched a slightly newer ffmpeg version from Nov 2020, see here; hopefully, current ffmpeg 5.x.x could be patched (and compiled?) in a similar fashion ...
>
> Source: SO
>
> Later addition:
> More recent patches & compilation guide here ...

There is no shortage of such patches out there; [one was even submitted to FFmpeg](https://patchwork.ffmpeg.org/project/ffmpeg/list/?series=4248). The problem is that these patches are all slightly different, and I'm not knowledgeable of codec internals to sort out the differences.

1. The patch from FFmpeg mailing list assigns codec ID of [10 to HEVC](https://patchwork.ffmpeg.org/project/ffmpeg/patch/20210701141528.776-1-jamrial@gmail.com/), [11 to VP8, and 12 to VP9](https://patchwork.ffmpeg.org/project/ffmpeg/patch/20210701141528.776-2-jamrial@gmail.com/):
   ```c
   FLV_CODECID_H265    = 10,
   FLV_CODECID_VP8     = 11,
   FLV_CODECID_VP9     = 12,
   ```

   [@runner365](https://github.com/runner365/ffmpeg_rtmp_h265/blob/5.1/flv.h) uses a totally different assignment (@xiaoxial patches match @runner365's HEVC and AV1):
   ```c
   FLV_CODECID_HEVC    = 12,
   FLV_CODECID_AV1     = 13,
   FLV_CODECID_VP8     = 14,
   FLV_CODECID_VP9     = 15,
   ```

   I've selected the value of 12 since that's what [mpegts.js](https://github.com/xqq/mpegts.js/blob/6bf10e8a5be96543aaf3fb4bb7dcfcf8e7a21f02/src/demux/flv-demuxer.js#L843) expects, but please be aware that someone out there streams __VP9__ over FLV using the codec ID of 12.

2. [FFmpeg mailing list patch](https://patchwork.ffmpeg.org/project/ffmpeg/patch/20210701141528.776-1-jamrial@gmail.com/) parses AVSTREAM HEADERS:
   ```c
   case FLV_CODECID_H265:
       par->codec_id = AV_CODEC_ID_H265;
       vstream->internal->need_parsing = AVSTREAM_PARSE_HEADERS;
       ret = 3;     // not 4, reading packet type will consume one byte
       break;
   ```

   @runner365 and @xiaoxial do not:
   ```c
   case FLV_CODECID_HEVC:
        par->codec_id = AV_CODEC_ID_HEVC;
        vstreami->need_parsing = AVSTREAM_PARSE_NONE;
        ret = 3;     // not 4, reading packet type will consume one byte
        break;
   ```

   (`AV_CODEC_ID_H265` and `AV_CODEC_ID_HEVC` are [synonyms](https://github.com/FFmpeg/FFmpeg/blob/3141dbb7adf1e2bd5b9ff700312d7732c958b8df/libavcodec/codec_id.h#L227)).

   @runner365 repeats the same block for VP8 and VP9, but [mailing list patch](https://patchwork.ffmpeg.org/project/ffmpeg/patch/20210701141528.776-2-jamrial@gmail.com/) does not. For VP8 and VP9, one byte is not consumed and `need_parsing` isn't touched at all:
   ```c
    case FLV_CODECID_VP8:
        par->codec_id = AV_CODEC_ID_VP8;
        break;
    case FLV_CODECID_VP9:
        par->codec_id = AV_CODEC_ID_VP9;
        break;
   ```

    I can't help but think that @runner365 simply copy-pasted the same block three times. That's why I've chosen to parse AVSTREAM HEADERS, but I may be wrong. Since the code is untested, a switch back to `AVSTREAM_PARSE_NONE` may end up being necessary.

3. OPUS supports only 48 kHz sample rates, which, unfortunately, is not reflected in [@runner365's flvenc.c](https://github.com/runner365/ffmpeg_rtmp_h265/blob/5.1/flvenc.c)
   ```c
   else if (par->codec_id == AV_CODEC_ID_OPUS) {
       return FLV_CODECID_OPUS | FLV_SAMPLERATE_44100HZ |
              FLV_SAMPLESSIZE_16BIT | FLV_STEREO;
   }
   ```

4. [@xiaoxial](https://github.com/VCDP/CDN/blob/master/FFmpeg_patches/0001-Add-SVT-HEVC-FLV-support-on-FFmpeg.patch) takes care to write a different HEVC end-of-stream:
   ```c
   if (par->codec_id == AV_CODEC_ID_H264 || par->codec_id == AV_CODEC_ID_MPEG4) {
       put_avc_eos_tag(pb, sc->last_ts);
   } else if (par->codec_id == AV_CODEC_ID_HEVC) {
       put_hevc_eos_tag(pb, sc->last_ts);
   }
   ```

   [@runner365](https://github.com/runner365/ffmpeg_rtmp_h265/blob/5.1/flvenc.c) does it neither for HEVC nor for VP8/9.

5. [@xiaoxial](https://github.com/VCDP/CDN/blob/master/FFmpeg_patches/0001-Add-SVT-HEVC-FLV-support-on-FFmpeg.patch), however, writes codec header to every frame: https://github.com/VCDP/CDN/issues/1.

6. [@runner365](https://github.com/runner365/ffmpeg_rtmp_h265/blob/5.1/flvenc.c) disallows missing PTS, [@xiaoxial](https://github.com/VCDP/CDN/blob/master/FFmpeg_patches/0001-Add-SVT-HEVC-FLV-support-on-FFmpeg.patch) does not:
   ```c
   if (par->codec_id == AV_CODEC_ID_H264 || par->codec_id == AV_CODEC_ID_MPEG4
       || par->codec_id == AV_CODEC_ID_H265 || par->codec_id == AV_CODEC_ID_VP8
       || par->codec_id == AV_CODEC_ID_VP9) {
         if (pkt->pts == AV_NOPTS_VALUE) {
             av_log(s, AV_LOG_ERROR, "Packet is missing PTS\n");
             return AVERROR(EINVAL);
   ```

   The check, however, [was originally added](https://github.com/FFmpeg/FFmpeg/commit/14d3384cf31a620ff451062f7263942f7fe3a972) to prevent [overflow](https://trac.ffmpeg.org/ticket/8152) in AVC NALU:
   ```c
   else if (par->codec_id == AV_CODEC_ID_H264 || par->codec_id == AV_CODEC_ID_MPEG4
            || par->codec_id == AV_CODEC_ID_HEVC || par->codec_id == AV_CODEC_ID_VP8
            || par->codec_id == AV_CODEC_ID_VP9) {
       avio_w8(pb, 1); // AVC NALU
       avio_wb24(pb, pkt->pts - pkt->dts);
   }
   ```

   If I'm not mistaken, NALU stands for [Network Abstraction Layer unit](https://en.wikipedia.org/wiki/Network_Abstraction_Layer), which

   > ... is a part of the H.264/AVC and HEVC video coding standards.

   Its applicability to VP8/VP9 is not at all obvious.

**Comment by pukkandan** [comment]
> * `yt-dlp` should probably warn users and remux HEVC-over-FLC upon encountering, to prevent playback issues.

Remuxing without being requested may not be a good idea. I think we should just leave the format deprioritized. Opinion, @Lesmiscore ?

**Comment by Lesmiscore** [comment]
> Remuxing without being requested may not be a good idea. I think we should just leave the format deprioritized. Opinion, @Lesmiscore ?

Agreed since we can't expect upstream to apply the patch

**Comment by nihil-admirari** [comment]
I've disabled [SSL for GMP](https://github.com/yt-dlp/FFmpeg-Builds/commit/31a0c86948748b688a9fd5796ffdfd6b8a643fbe). Builds are operational once again.

Having no access to real world HEVC-over-FLV, i've downloaded [an example](https://github.com/ossrs/srs/blob/develop/trunk/doc/source.200kbps.768x320.flv) from OSSRS, and recoded it as [per instructions](https://github.com/ossrs/srs/issues/465#ffmpeg-tools):
```sh
ffmpeg -re -i source.200kbps.768x320.flv -acodec copy -vcodec libx265 -f flv test.flv
```

Patched FFmpeg is capable of probing and playing the resulting file:
```
Input #0, flv, from 'test.flv':
  Metadata:
    major_brand     : isom
    minor_version   : 512
    compatible_brands: isomiso2avc1mp41
    encoder         : Lavf59.34.102
  Duration: 00:03:30.69, start: 0.040000, bitrate: 273 kb/s
  Stream #0:0: Video: hevc (Main), yuv420p(tv), 768x320 [SAR 1:1 DAR 12:5], 25 fps, 25 tbr, 1k tbn
  Stream #0:1: Audio: aac (LC), 44100 Hz, stereo, fltp, 30 kb/s
```

Non-patched FFmpeg throws:
```
[flv @ 0x7f612c000c80] Video codec (c) is not implemented. Update your FFmpeg version to the newest one from Git. If the problem still occurs, it means that your file has a feature which has not been implemented.
[flv @ 0x7f612c000c80] If you want to help, upload a sample of this file to https://streams.videolan.org/upload/ and contact the ffmpeg-devel mailing list. (ffmpeg-devel@ffmpeg.org)
[flv @ 0x7f612c000c80] Could not find codec parameters for stream 0 (Video: none ([12][0][0][0] / 0x000C), none): unknown codec
Consider increasing the value for the 'analyzeduration' (0) and 'probesize' (5000000) options
Input #0, flv, from 'test.flv':
  Metadata:
    major_brand     : isom
    minor_version   : 512
    compatible_brands: isomiso2avc1mp41
    encoder         : Lavf59.34.102
  Duration: 00:03:30.69, start: 0.000000, bitrate: 273 kb/s
  Stream #0:0: Video: none ([12][0][0][0] / 0x000C), none, 25 fps, 25 tbr, 1k tbn
  Stream #0:1: Audio: aac (LC), 44100 Hz, stereo, fltp, 30 kb/s
No decoder could be found for codec none
```

Nonetheless, testing on real-world examples would be greatly appreciated.

> Remuxing without being requested may not be a good idea.

At least a warning should be issued, when this format is selected, along the lines of “selected format is non-standard and unsupported by most player. Consider using --remux-video”.

**Comment by pukkandan** [comment]
> > Remuxing without being requested may not be a good idea.
> 
> At least a warning should be issued, when this format is selected, along the lines of “selected format is non-standard and unsupported by most player. Consider using --remux-video.”

We could add a note to `format_note`

**Comment by Lesmiscore** [comment]
> Nonetheless, testing on real-world examples would be greatly appreciated.

image

Looks okay to me (`-f source-1 --downloader ffmpeg https://live.bilibili.com/6`)

**Comment by gamer191** [comment]
Any plans to revert this, now that enhanced rtmp is standardised and supported by ffmpeg (see https://github.com/yt-dlp/FFmpeg-Builds/#1-support-for-hevc-over-flv)?

NOTE: I haven't tested if BiliBili actually follows this standard, which would allow it to be supported by ffmpeg

**Comment by pukkandan** [comment]
I'll leave it be for now.
